### ISSUE TITLE
testing infrastructure overhaul

### ISSUE THREAD

**Issue by grahamlopez** [QMCPACK/qmcpack#649]
An overhaul of the QMCPACK testing infrastructure could help improve coverage,
performance, and feedback for ongoing development activities. This issue is to start 
a discussion about possible approaches to this goal.

The current testing layout (both CI and nightlies) was covered at the Oct 2017
all-hands meeting and is available in the slides on Confluence. However, it
would be good to define an updated set of requirements before we start playing Tetris
and arranging various tests to available resources.

We should also be on the lookout for opportunities to economize and double count
on tests if/where possible to reduce redundancy and the number of combinations. This
could potentially let us add additional variety of hardware, software environments, test 
coverage, etc.

These lists are just to get us started. We will likely add/remove items from them.

## potential targets for tests

1. static checks for project policies e.g. code formatting, documentation. These
   should run very quickly and precede build tests. These shouldn't have any special
   architectural or environment requirements.
2. memory checking, runtime sanitizers. Ideally these would part of CI, but they
   might be slow so it may or may not be feasible. I think we can get away with
   only running these on one (or minimally few) architectures/environments.
3. hardware variants to consider and prioritize
    - x86_64
    - POWER9
    - NVIDIA (what kind of arch coverage? Kepler/Maxwell, Pascal, Volta
    - AVX512
    - BG/Q
    - others
4. software/environments to consider and prioritize
    - GNU compiler toolchain
    - LLVM compiler toolchain
    - Intel compiler toolchain
    - XL compiler toolchain
    - 3rd party libs or packages that we need to verify with?

## continuous integration tests

These tests provide confidence to the developer that submitted code hasn't
broken existing capabilities before a patch is merged into the codebase.

Some thoughts:

1. current unit tests likely fall into this category and they probably won't be
   gathering statistics
2. these tests are either completely reliable, or we provide some threshold
   that indicates a passing condition
3. we need to find a good compromise between covering as much code as possible,
   while running as quickly as possible
4. we want to cover as many targets (hardware and software environments) as
   possible, but to in order to manage turn around times, we probably won't be
   comprehensive
5. this category could also be used at installation time to provide confidence
   to the user that their build produced a ready-to-use executable

Besides tests to run and where to run them, some additional decision points for
CI testing:

1. what is the maximum time we are willing to wait for each patch to be tested?
2. we need to prioritize the hardware and software targets for CI (remember the
   remaining targets can likely be picked up in the nightlies)

## integration tests

These tests provide additional information about recently merged code by using
more comprehensive and expensive tests at regular intervals (e.g. nightly,
etc.).

Some thoughts:

1. current "short" and "long" tests likely fall into this category and can
   afford to gather statistics where necessary
2. these tests can take longer and will be run as some variation of "nightlies"
3. we might try to be more comprehensive in hardware/software coverage with
   this category of tests

Besides tests to run and where to run them, some additional decision points for
CI testing:

1. do we want to add performance regression testing to the mix? Likely some
   current correctness tests could be adapted or even serve double purpose, but
   we might need to make adjustments for getting performance information
   reliable enough for this purpose.

**Comment by grahamlopez** [comment]
Ye and I sat down Tuesday morning for an hour to discuss this, and here are some of our notes from that discussion.

## for small (CI/install) tests
- keep unit tests we have at present
- Ye likes to do one solid, one molecule
    - molecule only works with the 'real' code path - no specific suggestion at present
    - for the solid `diamondC_2x1x1_pp-vmc_sj-1-16` takes ~8seconds
        - real, complex
- `-R short` are too much for this category
-  CI is currently doing CPU 4 variants, GPU 4 variants (2 jobs)
    - we could split up into multiple jobs in an effort to reduce latency
- keep just checking gcc for CI. chances of other failures is small enough to
  wait for nightlies
- consider trading oxygen/rhea roles (oxygen for faster CI builds, rhea for
  longer nightly tests) in an effort to reduce CI latency
    - consider the GPU architecture version we want for CI
- we need a plan to migrate to SoA only for CI, deprecate AoS
    - do mixed variants for SoA, full for AoS
    - MP+SoA is likely the  "way of the future," very conservative users doing AoS+FP anyway
- architectures: keep x86, x86+GPU. we need add power + GPU
    - prioritize CPU POWER, but nice to have GPU POWER as well if possible
    - clang or XL? both are unreliable, choose the faster-compiling one
- provide an easier way to run CI tests locally during development (before
  pushing and submitting a PR)
    - maybe a script or a checklist for before pushing

## longer testing (nightlies, integration, )
- goals:
    - for each architecture, thoroughly test every code path
    - but not every version of software stack for every code path
    - be most thorough for newest possible versions of software stack
        - how do we prioritize which older versions to test?
- priority architectures
    - x86
    - x86+gpu
    - power
    - power+gpu
    - knl
    - bg/q (doesn't follow all-test rule)
- nice to have architectures
    - arm
    - amd
- short tests don't care about statistics and have a hard upper time limit on a
  given platform, but still need to perform all result checking (not involving
  stats)
- long tests are however long enough to get "sufficient statistics" for
  checking, even if it doesn't run a long time
    - maybe change the name away from "long tests" to get people in the right
      mindset as to what they are for (and not necessarily run a long time)

Next steps:
- Graham will come up with suggestions for the test slots (mapping of architectures, stack versions, code paths,  etc.) and Ye will come up with suggestions for the appropriate tests to fill the slots

## performance regression testing (new category - these are separate from the integration tests used for correctness checking)
- (we didn't get to this topic yet)

**Comment by grahamlopez** [comment]
Here is a [google spreadsheet](https://docs.google.com/spreadsheets/d/100wsO7vQCuAUzs1P2rSUuYi4z_KchYS3s34d9kt__iw/edit?usp=sharing) with a couple of tabs, one trying to work out allocation of testing resources and architecture/software stack coverage for CI and nightlies, the other for available qmcpack tests and their code path coverage and expense characteristics.

Anyone is welcome to participate with the sheet and/or make continue discussion in this thread.

**Comment by PDoakORNL** [comment]
I can push to CDash from pycicle a userspace CI engine I've been using on another project and my own private repos. I'd argue if you don't have CI on a configuration you don't actually support it. This allows any contributor to easily set up any build they care about and push results to the main CDash or their own.  Same idea as on your spread sheet, every PR gets tested. 

At the very least I'd like to set this up for myself. Manual regressions have no place in a modern fast moving project. This will require putting a CTestConfig.cmake somewhere so that pycicle knows where to submit results to and some other repo specific settings.

**Comment by prckent** [comment]
@PDoakORNL there will be a CI meeting soon where we can discuss this.

**Comment by grahamlopez** [comment]
@PDoakORNL You'll find no arguments from me on most of your points. We need to be intentional about what we're doing and support a sufficient variety of use cases. I'm looking forward to discussing this further with someone who has opinions about it. :)

**Comment by jtkrogel** [comment]
We have a developed testing strategy.
